File: eventloop/docs/tournament/timer_workload_digest_test.py
#!/usr/bin/env python3
"""Cross-language regression tests for timer workload digest framing."""

import hashlib
import json
from pathlib import Path
import re
import struct
import unittest


VECTOR_PATH = (
    Path(__file__).resolve().parents[2]
    / 'internal'
    / 'tournament'
    / 'testdata'
    / 'timerworkloaddigestvectors.json'
)
DOMAIN = 'go-utilpkg/eventloop/tournament/timer-parameters/v1'
LOWER_SHA256 = re.compile(r'^[0-9a-f]{64}$')


def framed_sha256(domain, fields):
    digest = hashlib.sha256()
    for field in [domain, *fields]:
        payload = field.encode('utf-8')
        digest.update(struct.pack('>Q', len(payload)))
        digest.update(payload)
    return digest.hexdigest()


class TimerWorkloadDigestTest(unittest.TestCase):
    def test_go_vectors_match_independent_python_framing(self):
        file = json.loads(VECTOR_PATH.read_text(encoding='utf-8'))
        self.assertEqual(
            set(file),
            {'schema_version', 'domain', 'framing_vectors', 'workload_vectors'},
        )
        self.assertEqual(file['schema_version'], 1)
        self.assertEqual(file['domain'], DOMAIN)

        framing_names = set()
        for vector in file['framing_vectors']:
            with self.subTest(framing=vector.get('name')):
                self.assertEqual(set(vector), {'name', 'domain', 'fields', 'sha256'})
                self.assertNotIn(vector['name'], framing_names)
                framing_names.add(vector['name'])
                self.assertRegex(vector['sha256'], LOWER_SHA256)
                self.assertEqual(
                    framed_sha256(vector['domain'], vector['fields']),
                    vector['sha256'],
                )
        self.assertEqual(len(framing_names), 3)

        workload_keys = set()
        for vector in file['workload_vectors']:
            with self.subTest(kind=vector.get('kind'), workload=vector.get('id')):
                self.assertEqual(
                    set(vector),
                    {'kind', 'id', 'parameter_type', 'parameter_json', 'sha256'},
                )
                key = (vector['kind'], vector['id'])
                self.assertNotIn(key, workload_keys)
                workload_keys.add(key)
                self.assertIn(vector['kind'], {'storage', 'qualification'})
                json.loads(vector['parameter_json'])
                self.assertRegex(vector['sha256'], LOWER_SHA256)
                self.assertEqual(
                    framed_sha256(
                        DOMAIN,
                        [
                            vector['kind'],
                            vector['id'],
                            vector['parameter_type'],
                            vector['parameter_json'],
                        ],
                    ),
                    vector['sha256'],
                )
        self.assertEqual(len(workload_keys), 26)


if __name__ == '__main__':
    unittest.main()

File: eventloop/internal/tournament/timerworkloaddigest_test.go
package tournament

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"reflect"
	"testing"
)

type timerWorkloadDigestVectorFile struct {
	SchemaVersion   uint8                        `json:"schema_version"`
	Domain          string                       `json:"domain"`
	FramingVectors  []timerWorkloadFramingVector `json:"framing_vectors"`
	WorkloadVectors []timerWorkloadDigestVector  `json:"workload_vectors"`
}

type timerWorkloadFramingVector struct {
	Name   string   `json:"name"`
	Domain string   `json:"domain"`
	Fields []string `json:"fields"`
	SHA256 string   `json:"sha256"`
}

type timerWorkloadDigestVector struct {
	Kind          string `json:"kind"`
	ID            string `json:"id"`
	ParameterType string `json:"parameter_type"`
	ParameterJSON string `json:"parameter_json"`
	SHA256        string `json:"sha256"`
}

func TestTimerWorkloadDigestVectors(t *testing.T) {
	want := canonicalTimerWorkloadDigestVectors(t)
	payload, err := os.ReadFile("testdata/timerworkloaddigestvectors.json")
	if err != nil {
		t.Fatal(err)
	}
	decoder := json.NewDecoder(bytes.NewReader(payload))
	decoder.DisallowUnknownFields()
	var got timerWorkloadDigestVectorFile
	if err := decoder.Decode(&got); err != nil {
		t.Fatal(err)
	}
	var trailing any
	if err := decoder.Decode(&trailing); err != io.EOF {
		t.Fatalf("timer workload digest vectors have trailing JSON: %v", err)
	}
	if reflect.DeepEqual(got, want) {
		return
	}
	canonical, err := json.MarshalIndent(want, "", "  ")
	if err != nil {
		t.Fatal(err)
	}
	t.Fatalf("timer workload digest vectors differ from canonical definitions; replacement:\n%s", canonical)
}

func canonicalTimerWorkloadDigestVectors(t *testing.T) timerWorkloadDigestVectorFile {
	t.Helper()
	file := timerWorkloadDigestVectorFile{
		SchemaVersion: 1,
		Domain:        timerParameterDigestDomain,
		FramingVectors: []timerWorkloadFramingVector{
			newTimerWorkloadFramingVector("unicode-empty-nul", "go-utilpkg/eventloop/tournament/framing-test/v1", "", "µs", "\x00"),
			newTimerWorkloadFramingVector("concatenation-a-bc", "a", "bc"),
			newTimerWorkloadFramingVector("concatenation-ab-c", "ab", "c"),
		},
	}
	seen := make(map[string]struct{}, timerStorageWorkloadCount+timerQualificationWorkloadCount)
	for _, definition := range timerStorageDefinitions() {
		file.WorkloadVectors = append(file.WorkloadVectors, newTimerWorkloadDigestVector(t, "storage", string(definition.ID), definition.Parameters, definition.ParameterSHA256, seen))
	}
	for _, definition := range timerQualificationDefinitions() {
		file.WorkloadVectors = append(file.WorkloadVectors, newTimerWorkloadDigestVector(t, "qualification", string(definition.ID), definition.Parameters, definition.ParameterSHA256, seen))
	}
	if len(file.WorkloadVectors) != timerStorageWorkloadCount+timerQualificationWorkloadCount {
		t.Fatalf("timer workload digest vector count = %d", len(file.WorkloadVectors))
	}
	return file
}

func newTimerWorkloadFramingVector(name, domain string, fields ...string) timerWorkloadFramingVector {
	return timerWorkloadFramingVector{
		Name: name, Domain: domain, Fields: fields,
		SHA256: fmt.Sprintf("%x", framedTimerSeal(domain, fields...)),
	}
}

func newTimerWorkloadDigestVector(t *testing.T, kind, id string, parameters any, digest string, seen map[string]struct{}) timerWorkloadDigestVector {
	t.Helper()
	key := kind + "\x00" + id
	if _, ok := seen[key]; ok {
		t.Fatalf("duplicate timer workload digest vector %q/%q", kind, id)
	}
	seen[key] = struct{}{}
	payload, err := json.Marshal(parameters)
	if err != nil {
		t.Fatal(err)
	}
	parameterType := reflect.TypeOf(parameters).String()
	if got := fmt.Sprintf("%x", framedTimerSeal(timerParameterDigestDomain, kind, id, parameterType, string(payload))); got != digest {
		t.Fatalf("timer workload digest %q/%q = %s, want %s", kind, id, got, digest)
	}
	return timerWorkloadDigestVector{
		Kind: kind, ID: id, ParameterType: parameterType,
		ParameterJSON: string(payload), SHA256: digest,
	}
}
